rfcs/framework-0040-relations.md

```diff
@@ -0,0 +1,321 @@
+
+In terms of Invenio the use cases are many. For instance a bibliographic record could be related to:
+
+- Resource type
+- Subjects/keywords
+- Languages
+- Authors
+- Organistaions
+- Geonames
+- ...
+
+## Previous work
+
+Invenio v1 had the concept of authority records. Authority records would be able to store all types of above examples in the same model.
+
+Early on in Invenio v2 and v3, there was discussions on an [Invenio-Authorities](https://github.com/inveniosoftware-attic/invenio-authorities/issues/1) module.
+
+Invenio v3 today solves part of the record relations via JSONRef support. For instance, a record can include a reference:
+
+```json
+{
+    "grant": {"$ref": "https://localhost/api/grants/..."},
+}
+```
+
+During e.g. indexing, the ``{"$ref": "..."}`` part is replaced with the related record. See also the [training exercise](https://github.com/inveniosoftware/training/tree/v3.1/11-linking-records).
+
+The issues with this approach are:
+
+- Technically diffcult to understand and debug because of the resolving magic happening behind the scenes. A special library, catches requests to special URLs, and translates them into database queries. This is non-intuitive and very difficult to trace when it's not working as expected.
+- Lack of fine grained control. It is not easily possible to control which fields from the referenced record is included. For instance, a biblographic record relates to a grant record that relates to a funder record. All three records will be dereferenced, causing the index JSON to become very big.
+- Security/performance - because JSONRef is based on making HTTP requests, there's a risk that badly formed URLs, will cause HTTP requests to be made. At best this will only cause performance issues, at worst it opens a backdoor for injecting content in the JSON document. While this could also be considered a feature, ability to relate to externally hosted records, we think the performance/security issues makes this technically impossible.
+
+
+## Design
+
+### Challenges
+
+The primary challenges of record relationships includes:
+
+- **Syntax**: How should a relation be made between two records?
+- **Integrity**: How is integrity checked and validated?
+- **Dereferencing**: When and how is a record dereferenced?
+- **Denormalization**: When a dereferenced record is updated, how is the related index updated?
+
+Each of above challenges is addressed in the following design.
+
+#### Syntax
+
+We propose the following syntax:
+
+```
+{
+    "<field>": {"id": "<persistent id>"},
+    "<field>": [{"id": "<persistent id>"}, ...]
+}
+```
+
+A relation is an object with a single key ``id`` with a string value for a persistent identifier for the related record.
+
+A relation is considered dereferenced if it contains more than one key.
```

Comment:
I thought the idea was that extensions could be placed in the relation too without necessarily meaning that the relation is dereferenced. e.g.,

undereferenced
```json
{
    "id": "identifierA",
    "ext": {
        "added_field1": "foo",
        "added_field2": "bar"
    }
}

undereferenced

```json
{
    "id": "identifierA",
    "key1": "val1",
    "ext": {
        "added_field1": "foo",
        "added_field2": "bar"
    }
}
```

Granted there could be dereferencing collisions, but that could be solved with a namespaced dereferencing:

new undereferenced

```json
{
    "id": "identifierA",
    "deref": {
        "key1": "val1",
    },
    "ext": {
        "added_field1": "foo",
        "added_field2": "bar"
    }
}
```

Reply:
We went away from it to keep it as simple as possible.

We were looking at solutions of using e.g. ``$id`` instead of ``id`` to know if a record was dereferenced, but then you would end up with that ``record.grant.id`` would not work when the record was not dereferenced (you would have to use ``record.grant.$id`` instead. (e.g. in a jinja template)